### PR TITLE
Sidebar: Avoid focus loss on active tab change

### PR DESCRIPTION
**Pull request by aduth**
Partially addresses: #8079

This pull request seeks to resolve an issue of focus loss when changing active sidebar tabs for the editor settings sidebar ("Document", "Block").

~Currently, this pull request includes all _but_ the fix itself. Notably, included is an end-to-end test case covering the expected behavior, and a global test event handler protecting against unexpected focus loss.~

**Update:** It has been implemented with https://github.com/WordPress/gutenberg/pull/10917/commits/72569ab71a79dcc777f58a1dffe8db4a9682fef6.

**Implementation notes:**

Focus loss occurs because each specific sidebar renders the entirety of the `.edit-post-wrapper` element. For editor settings, this also includes the tab header as well.

The bug occurs because React is unable to reconcile the two element trees, and is therefore forced to unmount and remount an entirely new sidebar wrapper.

An initial attempt here was to invert rendering, having each specific sidebar populating only the `Fill` of the sidebar component, whereas the `Slot` would be responsible for rendering the `.edit-post-wrapper` wrapper element. This has the added advantage of fixing an issue with the application of `withFocusReturn`, as currently this is broken since it's applied to the `Fill` (non-DOM) element. An issue with this approach is in respecting the `aria-label` application provided by each specific sidebar.

**Testing instructions:**

Verify that no focus loss occurs when changing tabs in the sidebar.

1. Navigate to Posts > Add New
2. Click "Document"
3. Press <kbd>Tab</kbd>
4. Press <kbd>Spacebar</kbd>
5. Note that focus remains on the "Block" tab after activation

Ensure end-to-end tests pass:

```
npm run test-e2e
```

### CODE REVIEW

**Comment by gziolo** [comment]
Thanks @aduth, this is very helpful. I will try to find some time today to start working on a fix.

**Comment by gziolo** [comment]
I consolidate both sidebars into one component with 678819 and it solves the issue. It reuses the same Fill with the updated implementation so it is able to re-render (instead of unmount and mount) even when the active tab changes.

**Comment by gziolo** [comment]
@vindl - just letting you know that this changes a bit the logic inside the `edit-post`'s `Layout` component.

**Comment by gziolo** [comment]
@afercia this should be ready to verify 👍

E2e tests fall but this is unrelated. I bet focus loss check should be enabled only for tests which use keyboard navigation, not to all of them.

**Comment by afercia** [comment]
@gziolo thanks. I'm not seeing the sidebar at all. When I click the Settings button, I see the `is-sidebar-open` CSS class gets applied but nothing is rendered, no sidebar. What I'm doing wrong?

File: test/e2e/support/utils.js [hunk 1]
```diff
@@ -459,3 +459,19 @@ export async function getAllBlocks() {
 		return select( 'core/editor' ).getBlocks();
 	} );
 }
+
+/**
+ * Binds to the document on page load which throws an error if a `focusout`
+ * event occurs without a related target (i.e. focus loss).
+ */
+export function observeFocusLoss() {
```
Comment (gziolo) [inline test/e2e/support/utils.js]: @aduth - it seems to fail too many test suites. I moved it to utils and applied only for the sidebar. Let's open a follow-up and this assertion where it fits.

**Comment by gziolo** [comment]
@afercia it was my mistake, it should be fixed with https://github.com/WordPress/gutenberg/pull/10917/commits/d5ff17263465d6dcac8d3773c3894c69156667b1#diff-b006c664acf96e390063ef30c215b267R60 - I did a typo when refactoring code at the end of the day ...

**Comment by afercia** [comment]
@gziolo LGTM 👍 thanks!

**Review by tofumatt — APPROVED** [review]
Ran the test steps locally and it worked a treat. The code makes sense and the tests passed, so :shipit:

File: packages/edit-post/src/components/sidebar/block-sidebar/index.js [hunk 2]
```diff
@@ -1,30 +0,0 @@
...
-const SIDEBAR_NAME = 'edit-post/block';
-
-const BlockSidebar = () => (
-	<Sidebar
-		name={ SIDEBAR_NAME }
-		label={ __( 'Editor settings' ) }
-	>
-		<SettingsHeader sidebarName={ SIDEBAR_NAME } />
-		<Panel>
-			<PanelBody className="edit-post-block-sidebar__panel">
```
Comment (aduth) [inline packages/edit-post/src/components/sidebar/block-sidebar/index.js]: There's still an occurrence of this now-removed class name in tests:

https://github.com/WordPress/gutenberg/blob/dd7e54bdd4bf15f919eda88b9ff88a836ae88c1a/test/e2e/specs/block-hierarchy-navigation.test.js#L35

Tests are currently running, but if they're passing that would be equally concerning.

**Comment by tofumatt** [comment]
Update: I ran the tests on master by mistake but I verified the fix manually on the branch, sorry. Running them on the right branch now 😓

**Review by tofumatt — DISMISSED** [review]
Locally I'm getting a failure on:

```
>   ● splitting and merging blocks › should remove at most one paragraph in forward direction
```

Sorry, I thought I clicked "submit review" hours ago but I guess not, I'm just the worst on this PR 😓

From the looks the snapshot is wrong so something is off here and needs fixing.

File: test/e2e/specs/block-hierarchy-navigation.test.js [hunk 3]
```diff
@@ -32,7 +32,7 @@ describe( 'Navigating the block hierarchy', () => {
 		await columnsBlockMenuItem.click();
 
 		// Tweak the columns count.
-		await page.focus( '.edit-post-block-sidebar__panel .components-range-control__number[aria-label="Columns"]' );
+		await page.focus( '.edit-post-settings-sidebar__panel-block .components-range-control__number[aria-label="Columns"]' );
```
Comment (gziolo) [inline test/e2e/specs/block-hierarchy-navigation.test.js]: Good catch, thanks 👍

File: test/e2e/support/utils.js (hunk 1 above)
Comment (gziolo) [reply test/e2e/support/utils.js]: I also reverted changes applied to `setup-test-framework.js` with https://github.com/WordPress/gutenberg/pull/10917/commits/46a078f397010bb57d8a1049a430bf23ebc6f8ac.

There is this issue that `afterEach` removes page listener but `afterAll` needs them, too.

**Comment by gziolo** [comment]
>   ● splitting and merging blocks › should remove at most one paragraph in forward direction

@tofumatt - it works properly for me locally and on Travis

**Comment by gziolo** [comment]
Travis is finally happy! 🎉

**Comment by aduth** [comment]
In spirit I give my approval, though despite @gziolo doing the bulk of the work, I can't review a pull request I'd opened 😄

**Review by youknowriad — APPROVED** [review]
despite @aduth doing the bulk of review work, he can't review the pull request he opened 😄so I'm approving :P

**Pull request merged**